pkg/kube/krt/nestedjoin.go

```diff
@@ -0,0 +1,282 @@
+			// After the first, safely merge into the result
+			for _, i := range objs {
+				key := GetKey(i)
+				if !found.InsertContains(key) {
+					// Only keep it if it is the first time we saw it, as our merging mechanism is to keep the first one
+					res = append(res, i)
+				}
+			}
+		}
+	}
+	return res
+}
+
+// nolint: unused // (not true, its to implement an interface)
+func (j *nestedjoin[T]) index(extract func(o T) []string) kclient.RawIndexer {
+	j.Lock() // Take a write lock since we're also updating the collection change handlers
+	defer j.Unlock()
+	ji := &dynamicJoinIndexer{indexers: make(map[collectionUID]kclient.RawIndexer)}
+	for _, c := range j.collections.List() {
+		ic := c.(internalCollection[T])
+		ji.indexers[ic.uid()] = ic.index(extract)
+	}
+	j.registerCollectionChangeHandlerLocked(func(e collectionChangeEvent[T]) {
+		ji.Lock()
+		defer ji.Unlock()
+		switch e.eventType {
+		case collectionMembershipEventAdd:
+			ji.indexers[e.collectionValue.uid()] = e.collectionValue.index(extract)
+		case collectionMembershipEventDelete:
+			delete(ji.indexers, e.collectionValue.uid())
+		}
+	})
+	return ji
+}
+
+func (j *nestedjoin[T]) Synced() Syncer {
+	return channelSyncer{
+		name:   j.collectionName,
+		synced: j.synced,
+	}
+}
+
+func (j *nestedjoin[T]) HasSynced() bool {
+	return j.syncer.HasSynced()
+}
+
+func (j *nestedjoin[T]) WaitUntilSynced(stop <-chan struct{}) bool {
+	return j.syncer.WaitUntilSynced(stop)
+}
+
+func (j *nestedjoin[T]) Register(f func(o Event[T])) HandlerRegistration {
+	return registerHandlerAsBatched(j, f)
+}
+
+func (j *nestedjoin[T]) RegisterBatch(f func(o []Event[T]), runExistingState bool) HandlerRegistration {
+	j.Lock() // Take a write lock since we're also updating the collection change handlers
+	defer j.Unlock()
+	syncers := make(map[collectionUID]Syncer)
+	removes := map[collectionUID]func(){}
+	for _, c := range j.collections.List() {
+		reg := c.RegisterBatch(f, runExistingState)
+		ic := c.(internalCollection[T])
+		removes[ic.uid()] = reg.UnregisterHandler
+		syncers[ic.uid()] = reg
+	}
+	djhr := &dynamicJoinHandlerRegistration{
+		syncers: syncers,
+		removes: removes,
+	}
+
+	j.registerCollectionChangeHandlerLocked(func(e collectionChangeEvent[T]) {
+		djhr.Lock()
+		defer djhr.Unlock()
+		switch e.eventType {
+		case collectionMembershipEventAdd:
+			reg := e.collectionValue.RegisterBatch(f, runExistingState)
+			djhr.removes[e.collectionValue.uid()] = reg.UnregisterHandler
+			djhr.syncers[e.collectionValue.uid()] = reg
+		case collectionMembershipEventDelete:
+			// Unregister the handler for this collection
+			remover := djhr.removes[e.collectionValue.uid()]
+			if remover == nil {
+				log.Warnf("Collection %v not found in %v", e.collectionValue.uid(), j.name())
+				return
+			}
+			remover()
+			delete(djhr.removes, e.collectionValue.uid())
+			delete(djhr.syncers, e.collectionValue.uid())
+
+			// Now send a final set of remove events for each object in the collection
```

Comment:
This may be overlapping with discussion in https://github.com/istio/istio/pull/55432/files. But if we have merging or overlapping keys across collections, this is going to not be quite right?

For example, if I have `[[k1],[k1]]` and delete the collection, I don't want to send a Delete(k1). Then we would have both deleted. 

I am not sure what the semantics of nestedjoin without a merge is exactly so hard to say precise what this should do. But in the merging world I would think you need to do a merge still. So this would not be a `Delete(k1)`

Reply:
Very good point; we'd need to do a merge to confirm that it's actually deleted and only then send a delete

Reply:
And send an update possibly

Reply:
I'll change this in #55432